ddtrace/debugging/_function/discovery.py

```diff
@@ -135,23 +136,80 @@ def _undecorate(f, name, path):
     # expected function name to guide the search and pick the correct function.
     # The recursion is needed in case of multiple decorators. We make it BFS
     # to find the function as soon as possible.
-    # DEV: We are deliberately not handling decorators that store the original
-    #      function in __wrapped__ for now.
-    if not (f.__code__.co_name == name and abspath(f.__code__.co_filename) == path):
-        seen_functions = {f}
-        q = deque([f])  # FIFO: use popleft and append
 
-        while q:
-            next_f = q.popleft()
-
-            for g in (
-                _.cell_contents for _ in (next_f.__closure__ or []) if _isinstance(_.cell_contents, FunctionType)
-            ):
-                if g.__code__.co_name == name and abspath(g.__code__.co_filename) == path:
-                    return g
-                if g not in seen_functions:
-                    q.append(g)
-                    seen_functions.add(g)
+    def match(g):
+        return g.__code__.co_name == name and abspath(g.__code__.co_filename) == path
+
+    if _isinstance(f, FunctionType) and match(f):
+        return f
+
+    seen_functions = {f}
+    q = deque([f])  # FIFO: use popleft and append
+
+    while q:
+        g = q.popleft()
+
+        # Look for a wrapped function
+        for attr in ("__wrapped__", "func"):
+            try:
+                wrapped = object.__getattribute__(g, attr)
+                if _isinstance(wrapped, FunctionType) and wrapped not in seen_functions:
```

Comment:
nit: this if logic is repeated multiple times. not sure if it would be cleaner to have in an helper function but there is also return in there... maybe there is a nice python way to do this. maybe not.

what I came up with is:
```
if isUnseenFunction(wrapped): 
    return wrapped if match(wrapped) else addToQueue(wrapped)
```
not sure if this any better.

Reply:
Since Python doesn't have inlining, I think it's best to be explicit and save a function call for simple expressions like this